File: Data Structures/Arrays/moveElementToEnd.js
// O(n) time | O(1) space
function moveElementToEnd(array, twoMove) {
  let leftPointer = 0;
  let rightPointer = array.length - 1;
  while (leftPointer < rightPointer) {
    while (leftPointer < rightPointer && array[rightPointer] === twoMove) {
      rightPointer--;
    }
    if (array[leftPointer] === twoMove) {
      [array[leftPointer], array[rightPointer]] = [
        array[rightPointer],
        array[leftPointer],
      ];
    }
    leftPointer++;
  }
  return array;
}

File: Data Structures/Arrays/moveElementToEnd.py
######### PROBLEM ##########
# You're given an array of integers and an integer. Write a function that moves
# all instances of that integer in the array to the end of the array and returns
# the array.
# The function should perform this in place (i.e., it should mutate the input
# array) and doesn't need to maintain the order of the other integers.

######### SOLUTION ##########
# Optimal Solution
# Time: O(n) where n is the length of the array
# Space: O(1)
def moveElementToEnd(array, toMove):
    # Write your code here.
    left = 0
    right = len(array) - 1

    while left < right:
        rightValue = array[right]
        leftValue = array[left]
        if rightValue == toMove:
            right = right - 1
            continue
        if leftValue != toMove:
            left = left + 1
            continue
        array[left], array[right] = rightValue, leftValue

    return array
